<a href="https://colab.research.google.com/github/KareemElshourbagy1/ML_Model/blob/main/ML_Task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
df = pd.read_csv("Churn_Modelling.csv")
df.head()

In [ ]:
df.shape
df.info()
df.isnull().sum()
df.describe()

In [ ]:
x=df.drop(columns=['RowNumber','CustomerId','Surname','Exited'])
y=df['Exited']
x.head()
y.head()

In [ ]:
x.dtypes

In [ ]:
from pandas.core.arrays import categorical
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
categorical_cols=['Geography','Gender']


In [ ]:
numeric_cols = [col for col in x.columns if col not in categorical_cols]

In [ ]:
preprocessor = ColumnTransformer(
    transformers=[
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_cols)
    ],
    remainder="passthrough"
)

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(
    max_depth=5,
    random_state=42
)

In [ ]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ("preprocessing", preprocessor),
    ("model", model)
])

In [ ]:
pipeline.fit(x_train, y_train)

In [ ]:
y_pred=pipeline.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("Accuracy:", accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

In [ ]:
print("Training Accuracy:", pipeline.score(x_train, y_train))
print("Testing Accuracy:", pipeline.score(x_test, y_test))

In [ ]:
from sklearn import tree
import matplotlib.pyplot as plt

plt.figure(figsize=(20, 10))

tree.plot_tree(
    pipeline.named_steps["model"],
    filled=True,
    max_depth=3
)

plt.show()

Churn_Modeling Work Flow


Churn Dataset
      ↓
Understand
      ↓
Remove useless columns
      ↓
x / y
      ↓
Categorical Encoding
      ↓
Train/Test Split
      ↓
DecisionTreeClassifier
      ↓
Training
      ↓
Prediction
      ↓
Evaluation
      ↓
Check Overfitting

Churn_Modeling finsh


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

bengaluru_house_prices start


In [33]:
house_df=pd.read_csv('bengaluru_house_prices.csv')
house_df.head()


,area_type,availability,location,size,society,total_sqft,bath,balcony,price
0,Super built-up Area,19-Dec,Electronic City Phase II,2 BHK,Coomee,1056,2.0,1.0,39.07
1,Plot Area,Ready To Move,Chikka Tirupathi,4 Bedroom,Theanmp,2600,5.0,3.0,120.00
2,Built-up Area,Ready To Move,Uttarahalli,3 BHK,NaN,1440,2.0,3.0,62.00
3,Super built-up Area,Ready To Move,Lingadheeranahalli,3 BHK,Soiewre,1521,3.0,1.0,95.00
4,Super built-up Area,Ready To Move,Kothanur,2 BHK,NaN,1200,2.0,1.0,51.00


In [34]:
house_df.shape
house_df.info()
house_df.isnull().sum()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


,0
area_type,0
availability,0
location,1
size,16
society,5502
total_sqft,0
bath,73
balcony,609
price,0


In [35]:
house_df.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

In [36]:
y_house=house_df['price']

In [37]:
x_house=house_df.drop(columns=['price'])


In [43]:
house_df.isnull().sum().sort_values(ascending=False)

,0
balcony,609
bath,73
size,16
location,1
area_type,0
availability,0
total_sqft,0
price,0


In [44]:
#house_df = house_df.drop(columns=["society"])
house_df["bhk"] = house_df["size"].str.extract("(\d+)").astype(float)

<>:2: SyntaxWarning: invalid escape sequence '\d'
<>:2: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipykernel_500/986561367.py:2: SyntaxWarning: invalid escape sequence '\d'
  house_df["bhk"] = house_df["size"].str.extract("(\d+)").astype(float)


In [45]:
house_df=house_df.dropna()

In [46]:
x_house = pd.get_dummies(
    x_house,
    columns=["location"],
    drop_first=True
)

In [50]:
def convert_sqft_to_num(x):
    try:
        if isinstance(x, str):
            if "-" in x:
                tokens = x.split("-")
                return (float(tokens[0]) + float(tokens[1])) / 2
            return float(x)
        return float(x)
    except:
        return None

In [51]:
house_df["total_sqft"] = house_df["total_sqft"].apply(convert_sqft_to_num)

/tmp/ipykernel_500/36383554.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  house_df["total_sqft"] = house_df["total_sqft"].apply(convert_sqft_to_num)


In [54]:
x_house = house_df[
    ["location", "total_sqft", "bath", "balcony", "bhk"]
]

y_house = house_df["price"]

In [55]:
x_house.isnull().sum()

,0
location,0
total_sqft,42
bath,0
balcony,0
bhk,0


In [56]:
x_house["total_sqft"] = x_house["total_sqft"].fillna(

    x_house["total_sqft"].median()
)

/tmp/ipykernel_500/1004126005.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_house["total_sqft"] = x_house["total_sqft"].fillna(


In [57]:
x_house.isnull().sum()

,0
location,0
total_sqft,0
bath,0
balcony,0
bhk,0


In [61]:
x_house = pd.get_dummies(
    x_house,
    columns=["location"],
    drop_first=True
)

In [62]:
x_house = x_house.astype(float)

In [64]:
print(x_house.dtypes)
print(x_house.shape)
print(x_house.isnull().sum().sum())

total_sqft                     float64
bath                           float64
balcony                        float64
bhk                            float64
location_ Banaswadi            float64
                                ...   
location_singapura paradise    float64
location_t.c palya             float64
location_tc.palya              float64
location_vinayakanagar         float64
location_whitefiled            float64
Length: 1268, dtype: object
(12710, 1268)
0


In [66]:
x_train_h, x_test_h, y_train_h, y_test_h = train_test_split(
    x_house,
    y_house,
    test_size=0.2,
    random_state=42
)

In [67]:
from sklearn.tree import DecisionTreeRegressor
house_model = DecisionTreeRegressor(
    max_depth=10,
    random_state=42
)

In [68]:
house_model.fit(x_train_h, y_train_h)

DecisionTreeRegressor(max_depth=10, random_state=42)

In [69]:
house_pred = house_model.predict(x_test_h)

In [70]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae = mean_absolute_error(y_test_h, house_pred)

mse = mean_squared_error(y_test_h, house_pred)

rmse = np.sqrt(mse)

r2 = r2_score(y_test_h, house_pred)

print("MAE:", mae)
print("RMSE:", rmse)
print("R²:", r2)

MAE: 35.04334121630886
RMSE: 85.91201775046476
R²: 0.6057225666770363


In [71]:
train_r2 = house_model.score(X_train_h, y_train_h)
test_r2 = house_model.score(X_test_h, y_test_h)

print("Training R²:", train_r2)
print("Testing R²:", test_r2)

Training R²: 0.8271449293852675
Testing R²: 0.6055350866536402


In [72]:
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeRegressor
tree = DecisionTreeRegressor(random_state=42)
param_grid = {
    "max_depth": [3, 5, 7, 10, 15, 20],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4]
}

In [73]:
grid_search = GridSearchCV(
    estimator=tree,
    param_grid=param_grid,
    cv=5,
    scoring="r2",
    n_jobs=-1
)

Currently, we have identified the best parameter.

In [75]:
grid_search.fit(x_train_h, y_train_h)

GridSearchCV(cv=5, estimator=DecisionTreeRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [3, 5, 7, 10, 15, 20],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10]},
             scoring='r2')

In [76]:
print(grid_search.best_params_)

{'max_depth': 5, 'min_samples_leaf': 4, 'min_samples_split': 10}


In [77]:
print(grid_search.best_score_)

0.5292312098666951


In [78]:
best_model = grid_search.best_estimator_
best_pred = best_model.predict(x_test_h)

In [79]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae = mean_absolute_error(y_test_h, best_pred)
rmse = np.sqrt(mean_squared_error(y_test_h, best_pred))
r2 = r2_score(y_test_h, best_pred)

print("MAE:", mae)
print("RMSE:", rmse)
print("R²:", r2)

MAE: 36.11318015075128
RMSE: 85.27229121128565
R²: 0.6115725200479711


We actually created a simplified version of the dataset. We used: location total_sqft bath balcony bhk However, the original dataset contains additional information such as: area_type availability society The problem might be that: The features we fed into the model aren't sufficient to predict the house price with higher accuracy.

 Garbage in → Garbage out...

 Because if the model has limited information, we can't expect excellent performance, no matter how much tuning we do.

Initial Model:
DecisionTreeRegressor

Initial Performance:
MAE = 35.04
RMSE = 85.91
R² = 0.605

After Hyperparameter Tuning:

MAE = 36.11
RMSE = 85.27
R² = 0.6115

Hyperparameter tuning resulted in a slight improvement in R² and RMSE, while MAE increased slightly. Overall, the improvement was limited.

Finally, I didn’t just train a model; I evaluated it, checked for overfitting, tuned its hyperparameters.